### PR TITLE
OSIDB-2306: Add validation function for embargoed and unembargoed dates

### PR DESCRIPTION
**Pull request by alizard0**
It implements [OSIDB-2306](https://issues.redhat.com/browse/OSIDB-2306)

### CODE REVIEW

File: src/components/FlawForm.vue [hunk 1]
```diff
@@ -84,7 +84,8 @@ const errors = {
   cwe_id: null,
   major_incident_state: null,
   reported_dt: null,
-  unembargo_dt: null,
+  embargo_old_public_dt: "Set the current or a future date for an embargoed flaw.",
+  unembargo_future_public_dt: "Set the current or an old date for unembargoed flaw.",
```
Comment (JakubFrejlach) [inline src/components/FlawForm.vue:88]: "Old date" sounds weird in this context, I would be for earlier/past/previous, wdyt?

File: src/components/FlawForm.vue [hunk 2]
```diff
@@ -97,6 +98,22 @@ const flawCvss3CaculatorLink = computed(
...
+  if (flaw.embargoed) {
+    // if embargoed and updated date is older than now, it shows an error to set the current date or a future date instead
+    if (DateTime.fromISO(flaw.unembargo_dt as string).toISODate() < DateTime.now().toISODate()) {
+      return errors.embargo_old_public_dt;
+    }
+  } else {
+    // if NOT embargoed and updated date is in the future, it shows an error, to set the current date or an older date instead
+    if (DateTime.fromISO(flaw.unembargo_dt as string).toISODate() > DateTime.now().toISODate()) {
+      return errors.unembargo_future_public_dt;
+    }
```
Comment (JakubFrejlach) [inline src/components/FlawForm.vue:112]: There is also one additional case which is apparently not mentioned in the Jira, but should be probably added. When Flaw is not embargoed and the unembargo_dt is not set, this is also invalid case and user should be prompted to set the unembargo_dt correctly. Also just a note, all these things are of course validated on the OSIDB side, so this servers merely as a convenience for user so he can observe these errors earlier.

**Review by JakubFrejlach — CHANGES_REQUESTED** [review]
Some suggestions. Also these changes should be rebased on top of the current main as linter should run on PR.

**Comment by JakubFrejlach** [comment]
Also I am not sure about the error message in this case:
image

Error message kinda blends in with the unembargo button, maybe we can adjust the colour a little bit? :thinking: Or add some visual borders?

File: src/components/FlawForm.vue (hunk 1 above)
Comment (alizard0) [reply src/components/FlawForm.vue:88]: Right, I agree.

File: src/components/FlawForm.vue (hunk 2 above)
Comment (alizard0) [reply src/components/FlawForm.vue]: What happens when is embargoed and dt is not set?

I have just pushed this commit:

https://github.com/RedHatProductSecurity/osim/pull/105/commits/4c26ecd4d451f2fd695dfc844f4a78281918831f#diff-bb5d5f91b338ff8b35ca76429e074f4af4afcd509d5adf9d6004016b419fd6cdR108

https://github.com/RedHatProductSecurity/osim/pull/105/commits/4c26ecd4d451f2fd695dfc844f4a78281918831f#diff-967a205363234f686e2f6df964ed2631bc6e26c2ba4f9d69abad87898d0d3461R443

Does it solves the use case?

File: src/components/FlawForm.vue [hunk 3]
```diff
@@ -101,6 +102,27 @@ const onReset = () => {
   flaw.value = deepCopyFromRaw(initialFlaw.value as Record<string, any>) as ZodFlawType;
 };
 
+const validateFlawEmbargoDates = computed(
+  () => {
+    let unembargo_dt = DateTime.fromISO(String(flaw.value.unembargo_dt)).toISODate();
```
Comment (tboyetteredhat) [inline src/components/FlawForm.vue:107]: Please use const instead of let.

**Review by tboyetteredhat — APPROVED** [review]

**Comment by tboyetteredhat** [comment]
@alizard0 Please fix test failures. https://github.com/RedHatProductSecurity/osim/actions/runs/8390059868/job/22977542203?pr=105#step:6:3983
image

File: src/components/__tests__/FlawForm.spec.ts [hunk 4]
```diff
@@ -9,10 +9,7 @@ import LabelEditable from '@/components/widgets/LabelEditable.vue';
 import IssueFieldStatus from '@/components/IssueFieldStatus.vue';
 import FlawForm from '../FlawForm.vue';
 import { useRouter } from 'vue-router';
-import LabelDiv from '../widgets/LabelDiv.vue';
-import LabelSelect from '../widgets/LabelSelect.vue';
```
Comment (tboyetteredhat) [inline src/components/__tests__/FlawForm.spec.ts:13]: I think these lines are needed.

**Review by tboyetteredhat — CHANGES_REQUESTED** [review]

File: src/components/__tests__/FlawForm.spec.ts (hunk 4 above)
Comment (alizard0) [reply src/components/__tests__/FlawForm.spec.ts:13]: It was a merge issue? Let me review it.

**Review by tboyetteredhat — APPROVED** [review]

File: src/components/FlawForm.vue [hunk 5]
```diff
@@ -87,7 +87,8 @@ const errors = {
   cwe_id: null,
   major_incident_state: null,
   reported_dt: null,
-  unembargo_dt: null,
+  embargo_old_public_dt: 'Set the current or a future date for an embargoed flaw.',
+  unembargo_future_public_dt: 'Set the current or an previous date for unembargoed flaw.',
```
Comment (C-Valen) [inline src/components/FlawForm.vue:91]: I think it would be simpler to return the strings from the validation method.

**Review by C-Valen — COMMENTED** [review]
IMO we should avoid overlapping validation messages:
image

I think it could be better something like  [ExampleValidatedForm.vue](https://github.com/RedHatProductSecurity/osim/blob/main/src/components/ExampleValidatedForm.vue)
image

Additionally not sure if the validation can be made with zod for this logic, in which case I'd say that could be preferred.

File: src/components/FlawForm.vue (hunk 2 above)
Comment (JakubFrejlach) [reply src/components/FlawForm.vue]: No, I don't think so.
Flaw is embargoed and unembargo_dt is not set -> valid state
Flaw is not embargoed and unembargo_dt is not set -> invalid state

**Review by JakubFrejlach — COMMENTED** [review]
Added some comments. Also I don't think it is necessary to include usernames in the commit names, it does not add much of a value and it clutters the git history.

File: src/components/FlawForm.vue (hunk 5 above)
Comment (alizard0) [reply src/components/FlawForm.vue:91]: I just defined them in the _const errors_ variable because I saw it was already defined.

**Comment by alizard0** [comment]
@JakubFrejlach I covered the missing use case.
@C-Valen I changed LabelEditable warning to match the LabelInput as you suggested.

**Comment by superbuggy** [comment]
Hey @alizard0,  #124 potentially covers the functionality here--the feature's scope evolved a bit to encompass this work. There we are using the `zod` library to perform the validations. We could huddle perhaps on Monday to look at how that works

**Comment by C-Valen** [comment]
> Hey @alizard0, #124 potentially covers the functionality here--the feature's scope evolved a bit to encompass this work. There we are using the `zod` library to perform the validations. We could huddle perhaps on Monday to look at how that works

May be works better tuesday (or any other day) I think monday is holiday on some countries.
I am also interested on that huddle for zod validations.

**Comment by alizard0** [comment]
Close PR because it overlaps with https://github.com/RedHatProductSecurity/osim/pull/124 and we agreed to merge the changes developed here into the other PR.

**Pull request closed without merging**